Client/Frontend/Browser/LoginsHelper.swift

```diff
@@ -180,48 +179,185 @@ class LoginsHelper: TabContentScript {
-        
-        profile.logins.getLoginsForProtectionSpace(login.protectionSpace).uponQueue(.main) { res in
-            var jsonObj = [String: Any]()
-            if let cursor = res.successValue {
-                log.debug("Found \(cursor.count) logins.")
-                jsonObj["requestId"] = requestId
-                jsonObj["name"] = "RemoteLogins:loginsFound"
-                jsonObj["logins"] = cursor.compactMap { loginData -> [String: String]? in
-                    if frameInfo.isMainFrame {
-                        return loginData?.toDict()
-                    }
-                    
-                    // The frame must belong to the same security origin
-                    if let currentHost = currentHost,
-                       !currentHost.isEmpty,
-                       currentHost == frameHost {
-                        // Prevent XSS on non main frame
-                        // If it is not the main frame, return username only, but no password!
-                        // Chromium does the same on iOS.
-                        // Firefox does NOT support third-party frames or iFrames.
-                        loginData?.update(password: "", username: loginData?.username ?? "")
-                        return loginData?.toDict()
+    
+    private func updateORSaveCredentials(for url: URL, script: [String: Any]) {
+        guard let scriptCredentials = passwordAPI.fetchFromScript(url, script: script),
+              let username = scriptCredentials.usernameValue,
+              scriptCredentials.usernameElement != nil,
+              let password = scriptCredentials.passwordValue,
+              scriptCredentials.passwordElement != nil else {
+            log.debug("Missing Credentials from script")
+            return
+        }
+        
+        if password.isEmpty {
+            log.debug("Empty Password")
+            return
+        }
+        
+        passwordAPI.getSavedLogins(for: url, formScheme: .typeHtml) { [weak self] logins in
+            guard let self = self else { return }
+            
+            for login in logins {
+                guard let usernameLogin = login.usernameValue else {
+                   continue
+                }
+                
+                if usernameLogin.caseInsensitivelyEqual(to: username) {
+                    guard password == login.passwordValue else {
+                        return
                     }
                     
-                    return nil
+                    self.showUpdatePrompt(from: login, to: scriptCredentials)
+                    return
+                } else {
+                    self.showAddPrompt(for: scriptCredentials)
+                    return
                 }
             }
+            
+            self.showAddPrompt(for: scriptCredentials)
+        }
+    }
+    
+    private func showAddPrompt(for login: PasswordForm) {
+        addSnackBarForPrompt(for: login, isUpdating: false) { [unowned self] in
```

Comment:
A bit scared of this `unowned` usage here (and below), the ownership chain here is not obvious.  `snackBar` is owned by `LoginsHelper` (self), but the snack bar itself is also owned by the `Tab` and then added as a subview to BVC? It is quite mirky and I am not sure that it is guaranteed that LoginsHelper will always be alive when the snack bar is? In general may be safer to use `weak` here.

Reply:
Fix at https://github.com/brave/brave-ios/pull/5013/commits/1a6d7cfd93a42bc6c37576fe5ddbca88aba7fa7c